src/Symfony/Component/Messenger/Transport/Doctrine/DoctrineReceiver.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Doctrine;
+
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\Serializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class DoctrineReceiver implements ReceiverInterface
+{
+    private $connection;
+    private $serializer;
+    private $shouldStop = false;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer = null)
+    {
+        $this->connection = $connection;
+        $this->serializer = $serializer ?? Serializer::create();
+    }
+
+    public function receive(callable $handler): void
+    {
+        while (!$this->shouldStop) {
+            $this->connection->requeueMessages();
+            $doctrineEnvelope = $this->connection->get();
+
+            if (null === $doctrineEnvelope) {
+                $handler(null);
+
+                usleep(200000);
+                if (\function_exists('pcntl_signal_dispatch')) {
+                    pcntl_signal_dispatch();
+                }
+
+                continue;
+            }
+
+            try {
+                $handler($this->serializer->decode([
+                    'body' => $doctrineEnvelope['body'],
+                    'headers' => $doctrineEnvelope['headers'],
+                ]));
+
+                $this->connection->ack($doctrineEnvelope['id']);
```

Comment:
We will be able to get rid of this after @weaverryan's PR about retry is over the line 🎉